File: python/posted/utils/definitions.py
import copy
import itertools
import warnings
from pathlib import Path
from typing import Literal

from posted.settings import default_currency
from posted.utils.read import read_yml_file


def read_definitions(directory_path: Path, flows: dict, techs: dict):
    assert directory_path.is_dir()

    # read all definitions and tags
    definitions = {}
    tags = {}
    for file_path in directory_path.rglob('*.yml'):
        if file_path.name.startswith('tag_'):
            tags |= read_yml_file(file_path)
        else:
            definitions |= read_yml_file(file_path)

    # read tags from flows and techs
    tags['Flow IDs'] = {
        flow_id: {}
        for flow_id, flow_specs in flows.items()
    }
    tags['Tech IDs'] = {
        tech_id: {key: val for key, val in tech_specs.items() if key in ['reference_flow_id', 'case_fields']}
        for tech_id, tech_specs in techs.items()
    }

    # insert tags
    for tag, items in tags.items():
        definitions = replace_tags(definitions, tag, items)

    # remove definitions where tags could not been replaced
    if any('{' in key for key in definitions):
        warnings.warn('Tokens could not be replaced correctly.')
        definitions = {k: v for k, v in definitions.items() if '{' not in k}

    # insert tokens
    tokens = {
        'default currency': lambda def_specs: default_currency,
    } | {
        f"default {subtype} flow unit {unit_component}": unit_token_func(subtype, unit_component, flows)
        for subtype in ('reported', 'reference')
        for unit_component in ('full', 'raw', 'variant')
    }
    for def_key, def_specs in definitions.items():
        for def_property, def_value in def_specs.items():
            for token_key, token_func in tokens.items():
                if f"{{{token_key}}}" in def_value:
                    def_specs[def_property] = def_specs[def_property].replace(f"{{{token_key}}}", token_func(def_specs))

    return definitions


def list_cases(case_fields: dict):
    cases = []
    for case_field_combinations in itertools.chain.from_iterable(itertools.combinations(case_fields.values(), r) for r in range(len(case_fields)+1)):
        cases.extend([' '.join(list(e)) for e in itertools.product(*case_field_combinations)])

    return cases


def replace_tags(definitions: dict, tag: str, items: dict[str]):
    definitions_with_replacements = {}
    for def_name, def_specs in definitions.items():
        if f"{{{tag}}}" not in def_name:
            definitions_with_replacements[def_name] = def_specs
        else:
            for item_name, item_specs in items.items():
                cases = [''] if 'case_fields' not in item_specs else list_cases(item_specs['case_fields'])
                for case_name in cases:
                    item_name_case = item_name if not case_name else f"{item_name} {case_name}"
                    item_desc = item_specs['description'] if 'description' in item_specs else item_name_case
                    def_name_new = def_name.replace(f"{{{tag}}}", item_name_case)
                    def_specs_new = copy.deepcopy(def_specs)
                    def_specs_new |= item_specs
                    def_specs_new['description'] = def_specs['description'].replace(f"{{{tag}}}", item_desc)
                    for k, v in def_specs_new.items():
                        if k == 'description' or not isinstance(v, str):
                            continue
                        def_specs_new[k] = v.replace(f"{{{tag}}}", item_name_case)
                    definitions_with_replacements[def_name_new] = def_specs_new

    return definitions_with_replacements


def unit_token_func(subtype: Literal['reported', 'reference'], unit_component: Literal['full', 'raw', 'variant'], flows: dict):
    return lambda def_specs: (
        'ERROR'
        if f"{subtype}_flow_id" not in def_specs or def_specs[f"{subtype}_flow_id"] not in flows else
        (
            flows[def_specs[f"{subtype}_flow_id"]]['default_unit']
            if unit_component == 'full' else
            flows[def_specs[f"{subtype}_flow_id"]]['default_unit'].split(';')[0]
            if unit_component == 'raw' else
            ';'.join([''] + flows[def_specs[f"{subtype}_flow_id"]]['default_unit'].split(';')[1:2])
            if unit_component == 'variant' else
            'UNKNOWN'
        )
    )

File: R/definitions.R
library(dplyr)
library(purrr)
library(tidyr)
library(stringr)

#' @title read_definitions
#'
#' @description Reads YAML files from definitions directory, extracts tags, inserts tags into
#' definitions, replaces tokens in definitions, and returns the updated definitions.
#'
#' @param definitions_dir Character. Path leading to the definitions.
#' @param flows List. Dictionary containing the different flow types. Each key represents a flow type, the corresponding
#'              value is a dictionary containing key value pairs of attributes like density, energy content and their
#'              values.
#' @param techs List. Dictionary containing information about different technologies. Each key in the
#'              dictionary represents a unique technology ID, and the corresponding value is a dictionary containing
#'              various specifications for that technology, like 'description', 'class', 'primary output' etc.
#'
#' @return List. Dictionary containing the definitions after processing and replacing tags and tokens.
#'
#' @export
read_definitions <- function(definitions_dir, flows, techs) {


  # chek that variables exists and is a directory
  if(!dir.exists(definitions_dir)) {
    stop(paste0("Should be a directory but is not: ", definitions_dir))
  }

  # read all definitions and tags
  definitions <- list()
  tags <- list()

  file_paths <- list.files(path = definitions_dir, pattern = "\\.yml$", recursive = TRUE, full.names = TRUE)
  for (file_path in file_paths) {
    if (grepl("^tag_", basename(file_path))) {
      tags <- c(tags, read_yml_file(file_path))
    } else {
      definitions <- c(definitions, read_yml_file(file_path))
    }
  }

  # read tags from flows and techs
  tags[['Flow IDs']] <- map(flows, ~list())
  tags[['Tech IDs']] <- map(techs, function(item) {item['primary_output']})

  # instert tags
  for (tag in names(tags)) {
    definitions <- replace_tags(definitions, tag, tags[[tag]])
  }

  # remove definitions where tags could not been replaced
  if (any(sapply(definitions, function(x) "\\{" %in% names(x)))) {
    warning('Tokens could not be replaced correctly.')
    definitions <- purrr::keep(definitions, ~ !any(sapply(.x, function(x) "{" %in% names(x))))
  }

  # insert tokens
  tokens <- list(
    'default currency' = function(def_specs) default_currency,
    'primary output' = function(def_specs) def_specs$primary_output
  ) %>%
   c(set_names(
        lapply(c('full', 'raw', 'variant'), function(unit_component) {
          unit_token_func(unit_component, flows)
        }),
        sprintf('default flow unit %s', c('full', 'raw', 'variant'))
      ))
  for (def_key in names(definitions)) {
    for (def_property in names(definitions[[def_key]])) {
      def_value <- definitions[[def_key]][[def_property]]

      for (token_key in names(tokens)) {
        token_func <- tokens[[token_key]]
        if (is.character(def_value) && grepl(paste0("\\{", token_key, "\\}"), def_value)) {
          definitions[[def_key]][[def_property]] <- gsub(paste0("\\{", token_key, "\\}"), token_func(definitions[[def_key]]), definitions[[def_key]][[def_property]])

      }
    }
  }
  }
  return(definitions)
}



#' @title replace_tags
#'
#' @description Replaces specified tags in dictionary keys and values with corresponding
#' items from another dictionary.
#'
#' @param definitions List. Dictionary containing the definitions, where the tags should be replaced by the items.
#' @param tag Character. String to identify where replacements should be made in the definitions. Specifies
#'            the placeholder that needs to be replaced with actual values from the \code{items} dictionary.
#' @param items List. Dictionary containing the items from which to replace the definitions.
#'
#' @return List. Dictionary containing the definitions with replacements based on the provided tag and items.
#'
#' @export
replace_tags <- function(definitions, tag, items) {


  definitions_with_replacements <- list()

  # Precompute regular expressions for tag replacement
  tag_regex <- paste0("\\{", tag, "\\}")

  for (def_name in names(definitions)) {
    def_specs <- definitions[[def_name]]

    if (grepl(tag_regex, def_name)) {
      for (item_name in names(items)) {
        item_specs <- items[[item_name]]
        item_desc <- ifelse('description' %in% names(item_specs), item_specs[['description']], item_name)

        def_name_new <- gsub( paste0("\\{", tag, "\\}"), item_name, def_name)
        def_specs_new <- c(def_specs, item_specs)
        def_specs_new <- def_specs_new[!duplicated(names(def_specs_new))]

        # Replace tags in description
        def_specs_new$description <- gsub( paste0("\\{", tag, "\\}"), item_desc, def_specs$description)

        # Replace tags in other specs
        def_specs_new <- lapply(def_specs_new, function(spec) {
          if (is.character(spec) && !identical(names(def_specs_new)[which(def_specs_new == spec)], 'description')) {
            spec <- gsub( paste0("\\{", tag, "\\}"), item_name, spec)
            spec <- gsub('\\{parent variable\\}', substr(def_name, 1, gregexpr( paste0("\\{", tag, "\\}"), def_name)[[1]] - 2), spec)
          }
          return(spec)
        })
        definitions_with_replacements[[def_name_new]] <- def_specs_new
      }
    } else {
      definitions_with_replacements[[def_name]] <- def_specs
    }
  }
  return(definitions_with_replacements)
}


#' @title unit_token_func
#'
#' @description Takes a unit component type and a dictionary of flows, and returns a lambda function
#' that extracts the default unit based on the specified component type from the flow
#' dictionary.
#'
#' @param unit_component Character. Specifies the type of unit token to be returned. Possible values are 'full', 'raw', 'variant'.
#' @param flows List. Dictionary containing the flows.
#'
#' @return Function. Lambda function that takes a dictionary \code{def_specs} as input. The lambda function
#' will return different values based on the \code{unit_component} parameter.
#'
#' @export
unit_token_func <- function(unit_component, flows) {
  return(function(def_specs) {
    if (!('flow_id' %in% names(def_specs)) || !(def_specs$flow_id %in% names(flows))) {
      return('ERROR')
    } else {
      if (unit_component == 'full') {
        return(flows[[def_specs$flow_id]]$default_unit)
      } else if (unit_component == 'raw') {
        return(strsplit(flows[[def_specs$flow_id]]$default_unit, ';')[[1]][1])
      } else if (unit_component == 'variant') {
        if (paste0(';', strsplit(flows[[def_specs$flow_id]]$default_unit, ';')[[1]][2])==";NA") {
          return ("")
        } else {
        return(paste0(';', strsplit(flows[[def_specs$flow_id]]$default_unit, ';')[[1]][2]))}
      } else {
        return('UNKNOWN')
      }
    }
  })
}
